modules/transactions/src/integrationTest/java/org/apache/ignite/internal/tx/readonly/ItReadOnlyTxAndLowWatermarkTest.java

```diff
@@ -0,0 +1,251 @@
+        builder.clusterConfiguration("ignite.gc.lowWatermark: {\n"
+                // Update frequently.
+                + "  updateInterval: 100\n"
+                + "}");
+    }
+
+    @BeforeEach
+    void createTable() {
+        node(0).sql().executeScript("CREATE TABLE " + TABLE_NAME + " (ID INT PRIMARY KEY, VAL VARCHAR)");
+    }
+
+    @ParameterizedTest
+    @EnumSource(TransactionalReader.class)
+    void roTransactionNoticesTupleVersionsMissingDueToGcOnDataNodes(TransactionalReader reader) throws Exception {
+        // TODO: remove the assumption when IGNITE-24119 is fixed.
+        assumeFalse(reader == TransactionalReader.MULTI_GET);
+
+        updateDataAvailabilityTimeToShortPeriod();
+
+        Ignite coordinator = node(0);
+        KeyValueView<Integer, String> kvView = kvView(coordinator);
+
+        insertOriginalValues(KEY_COUNT, kvView);
+
+        Transaction roTx = coordinator.transactions().begin(new TransactionOptions().readOnly(true));
+
+        updateToNewValues(KEY_COUNT, kvView);
+
+        waitTillLwmTriesToRaiseAndEraseOverrittenVersions();
+
+        IgniteException ex = assertThrows(IgniteException.class, () -> reader.read(coordinator, roTx));
+        assertThat(ex, isA(reader.sql() ? SqlException.class : TransactionException.class));
+        assertThat(ex, hasToString(containsString("Read timestamp is not available anymore.")));
+        assertThat("Wrong error code: " + ex.codeAsString(), ex.code(), is(Transactions.TX_STALE_READ_ONLY_OPERATION_ERR));
+    }
+
+    private void updateDataAvailabilityTimeToShortPeriod() {
+        IgniteImpl igniteImpl = unwrapIgniteImpl(node(0));
+
+        LowWatermarkConfiguration lwmConfig = igniteImpl.clusterConfiguration()
+                .getConfiguration(GcExtensionConfiguration.KEY)
+                .gc()
+                .lowWatermark();
+
+        assertThat(lwmConfig.dataAvailabilityTime().update(SHORT_DATA_AVAILABILITY_TIME_MS), willCompleteSuccessfully());
+    }
+
+    private static KeyValueView<Integer, String> kvView(Ignite coordinator) {
+        return coordinator.tables().table(TABLE_NAME).keyValueView(Integer.class, String.class);
+    }
+
+    private static void insertOriginalValues(int keyCount, KeyValueView<Integer, String> kvView) {
+        for (int i = 0; i < keyCount; i++) {
+            kvView.put(null, i, "original-" + i);
+        }
+    }
+
+    private static void updateToNewValues(int keyCount, KeyValueView<Integer, String> kvView) {
+        for (int i = 0; i < keyCount; i++) {
+            kvView.put(null, i, "updated-" + i);
+        }
+    }
+
+    private static void waitTillLwmTriesToRaiseAndEraseOverrittenVersions() throws InterruptedException {
+        Thread.sleep(2 * SHORT_DATA_AVAILABILITY_TIME_MS);
+    }
+
+    @CartesianTest
+    @WithSystemProperty(key = ResourceVacuumManager.RESOURCE_VACUUM_INTERVAL_MILLISECONDS_PROPERTY, value = "100")
+    void lwmIsAllowedToBeRaisedOnDataNodesAfterRoTransactionFinish(
```

Comment:
I'm curios how it's expected to work in case of abandoned RO transactions? I mean transactions that lost their coordinators. I believe that we should add test for such scenario.

Reply:
Added the corresponding test, it showed that this doesn't work, indeed. Implemented registration of transaction as a resource (analogous to cursors): on coordinator leave, the LWM is unlocked for this transaction